services-gateway-netty/src/main/java/io/scalecube/services/gateway/rsocket/RSocketGatewayAcceptor.java

```diff
@@ -22,96 +20,50 @@
   @Override
   public Mono<RSocket> accept(ConnectionSetupPayload setup, RSocket rsocket) {
     LOGGER.info("Accepted rsocket websocket: {}, connectionSetup: {}", rsocket, setup);
 
-    rsocket
-        .onClose()
-        .doOnTerminate(() -> LOGGER.info("Client disconnected: {}", rsocket))
-        .subscribe(null, th -> LOGGER.error("Exception on closing rsocket: {}", th.toString()));
-
     // Prepare message codec together with headers from metainfo
     HeadersCodec headersCodec = HeadersCodec.getInstance(setup.metadataMimeType());
     ServiceMessageCodec messageCodec = new ServiceMessageCodec(headersCodec);
-
-    return Mono.just(new GatewayRSocket(serviceCall, metrics, messageCodec));
-  }
-
-  /**
-   * Private extension class for rsocket. Holds gateway business logic in following methods: {@link
-   * #fireAndForget(Payload)}, {@link #requestResponse(Payload)}, {@link #requestStream(Payload)}
-   * and {@link #requestChannel(org.reactivestreams.Publisher)}.
-   */
-  private static class GatewayRSocket extends AbstractRSocket {
-
-    private final ServiceCall serviceCall;
-    private final GatewayMetrics metrics;
-    private final ServiceMessageCodec messageCodec;
-
-    /**
-     * Constructor for gateway rsocket.
-     *
-     * @param serviceCall service call coming from microservices.
-     * @param metrics gateway metrics.
-     * @param messageCodec message messageCodec.
-     */
-    private GatewayRSocket(
-        ServiceCall serviceCall, GatewayMetrics metrics, ServiceMessageCodec messageCodec) {
-      this.serviceCall = serviceCall;
-      this.metrics = metrics;
-      this.messageCodec = messageCodec;
-    }
-
-    @Override
-    public Mono<Void> fireAndForget(Payload payload) {
-      return Mono.defer(
-          () -> {
-            metrics.markRequest();
-            return serviceCall.oneWay(toMessage(payload));
-          });
-    }
-
-    @Override
-    public Mono<Payload> requestResponse(Payload payload) {
-      return Mono.defer(
-          () -> {
-            metrics.markRequest();
-            return serviceCall
-                .requestOne(toMessage(payload))
-                .map(this::toPayload)
-                .doOnNext(payload1 -> metrics.markServiceResponse());
-          });
-    }
-
-    @Override
-    public Flux<Payload> requestStream(Payload payload) {
-      return Flux.defer(
-          () -> {
-            metrics.markRequest();
-            return serviceCall
-                .requestMany(toMessage(payload))
-                .map(this::toPayload)
-                .doOnNext(payload1 -> metrics.markServiceResponse());
-          });
+    final RSocketGatewaySession resultRsocketSession =
+        new RSocketGatewaySession(
+            serviceCall, metrics, messageCodec, sessionEventHandler::mapMessage);
+
+    try {
+      sessionEventHandler.onSessionOpen(resultRsocketSession);
+    } catch (Exception e) {
+      return Mono.error(new ConnectionErrorException(e.getMessage()));
     }
```

Comment:
Why [block](https://github.com/scalecube/scalecube-gateway/pull/92/files#diff-2e591c9b063c48546b7e6484f3611f72R52-R56) is needed? Why not just `sessionHandler.onSessionOpen()` without wrapping into try-catch?

Reply:
I'm executing consumer's code here and want to explicitly wrap any possible exception from it into ConnectExc. 

*disclaimer that's borrowed from WebsocketGateway

Reply:
I'm executing consumer's code here and want to explicitly wrap any possible exception from it into ConnectExc.

*disclaimer that's borrowed from WebsocketGateway

Reply:
I think you mean this in `WebsocketGatewayAcceptor`: 

```java
  private Mono<Void> onConnect(WebsocketSession session) {
    try {
      gatewayHandler.onSessionOpen(session);
    } catch (Exception e) {
      return session.close(e.getMessage());
    }
  ...
```
Ok. Understood. It was done to close ws-session. But .. to say truth it's redundant. Just checked code in `reactor.netty.http.server.HttpServerHandle#onStateChange(63)`

Reply:
true, it's done already there.. will remove for both